In [3]:
import tensorflow as tf
from tensorflow.keras import models, layers

def build_car_part_classifier(input_shape=(224, 224, 3)):
    model = models.Sequential()
    
    model.add(layers.Input(shape=input_shape))
    
    model.add(layers.Conv2D(32, (3, 3), activation='relu'))
    model.add(layers.MaxPooling2D((2, 2)))
    
    model.add(layers.Conv2D(64, (3, 3), activation='relu'))
    model.add(layers.MaxPooling2D((2, 2)))
    
    model.add(layers.Conv2D(128, (3, 3), activation='relu'))
    model.add(layers.MaxPooling2D((2, 2)))
    
    model.add(layers.Flatten())
    
    model.add(layers.Dense(256, activation='relu'))
    
    model.add(layers.Dense(50, activation='softmax'))
    
    return model

In [4]:

cnn_model = build_car_part_classifier()
cnn_model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_3 (Conv2D)               │ (None, 222, 222, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_3 (MaxPooling2D)  │ (None, 111, 111, 32)   │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_4 (Conv2D)               │ (None, 109, 109, 64)   │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_4 (MaxPooling2D)  │ (None, 54, 54, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_5 (Conv2D)               │ (None, 52, 52, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_5 (MaxPooling2D)  │ (None, 26, 26, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_1 (Flatten)             │ (None, 86528)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 256)            │    22,151,424 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 50)             │        12,850 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 22,257,522 (84.91 MB)

 Trainable params: 22,257,522 (84.91 MB)

 Non-trainable params: 0 (0.00 B)

The custom Convolutional Neural Network employs a Sequential architecture designed for efficient hierarchical feature extraction. The network processes the input data through three consecutive blocks consisting of Conv2D and MaxPooling2D layers. We progressively increased the filter count (32, 64, and 128) across these blocks to allow the network to initially capture simple shapes and edges, and gradually learn highly complex visual patterns specific to car parts. Simultaneously, the max-pooling layers systematically downsample the spatial dimensions from 222x222 down to 26x26, filtering out noise and keeping the computational load manageable. After the Flatten layer transforms the 3D feature maps into a 1D vector of 86,528 elements, a 256-neuron Dense layer acts as the primary decision-making center. Finally, the architecture concludes with a 50-neuron Dense output layer using a softmax activation to output precise probabilities for each of the 50 car part classes.